src/core/reaction_ensemble.cpp

```diff
@@ -0,0 +1,2061 @@
+		if(current_reaction->len_product_types<current_reaction->len_reactant_types){
+			//hide superfluous reactant_types particles
+			for(int j=0;j<current_reaction->reactant_coefficients[i];j++){
+				int p_id ;
+				find_particle_type(current_reaction->reactant_types[i], &p_id);
+				*hidden_particles_properties=(double*) realloc(*hidden_particles_properties,sizeof(double)*(len_hidden_particles_properties+1)*number_of_saved_properties);
+				(*hidden_particles_properties)[len_hidden_particles_properties*number_of_saved_properties]=(double) p_id;
+				(*hidden_particles_properties)[len_hidden_particles_properties*number_of_saved_properties+1]= current_reaction_system.charges_of_types[find_index_of_type(current_reaction->reactant_types[i])];
+				(*hidden_particles_properties)[len_hidden_particles_properties*number_of_saved_properties+2]=(double) current_reaction->reactant_types[i];
+				len_hidden_particles_properties+=1;
+				hide_particle(p_id,current_reaction->reactant_types[i]);
+			}
+		} else {
+			//create additional product_types particles
+			for(int j=0;j<current_reaction->product_coefficients[i];j++){
+				int p_id = -1;
+				while (p_id == -1) {
+					p_id= create_particle(current_reaction->product_types[i]);
+				}
+				*p_ids_created_particles=(int*) realloc((*p_ids_created_particles),sizeof(int)*(len_p_ids_created_particles+1));
+				(*p_ids_created_particles)[len_p_ids_created_particles]=p_id;
+				len_p_ids_created_particles+=1;
+			}
+		}
+	}
+	*_len_changed_particles_properties=len_changed_particles_properties;
+	*_len_p_ids_created_particles=len_p_ids_created_particles;
+	*_len_hidden_particles_properties=len_hidden_particles_properties;
+}
+
+double calculate_factorial_expression(single_reaction* current_reaction, int* old_particle_numbers){
+	double factorial_expr=1.0;	
+	//factorial contribution of reactants
+	for(int i=0;i<current_reaction->len_reactant_types;i++) {
+		int nu_i=-1*current_reaction->reactant_coefficients[i];
+		int N_i0= old_particle_numbers[find_index_of_type(current_reaction->reactant_types[i])];
+		factorial_expr=factorial_expr*factorial_Ni0_divided_by_factorial_Ni0_plus_nu_i(N_i0,nu_i); //zeta = 1 (see smith paper) since we only perform one reaction at one call of the function
+	}
+	//factorial contribution of products
+	for(int i=0;i<current_reaction->len_product_types;i++) {
+		int nu_i=current_reaction->product_coefficients[i];
+		int N_i0= old_particle_numbers[find_index_of_type(current_reaction->product_types[i])];
+		factorial_expr=factorial_expr*factorial_Ni0_divided_by_factorial_Ni0_plus_nu_i(N_i0,nu_i); //zeta = 1 (see smith paper) since we only perform one reaction at one call of the function
+	}
+	return factorial_expr;
+}
+
+int generic_oneway_reaction(int reaction_id){
+	double volume = current_reaction_system.volume;
```

Comment:
Please make things that are not supposed to change anymore const, here and evrywhere